py4cast/datasets/poesy/__init__.py

```diff
@@ -72,121 +74,220 @@ def load_poesy_grid_info(name: str) -> GridConfig:
+        name = param.parameter_short_name
+        means = np.asarray(stats[name]["mean"])
+        std = np.asarray(stats[name]["std"])
 
-    def load_data(self, date: dt.datetime, term: List, member: int) -> np.array:
-        """
-        date : Date of file.
-        term : Position of leadtimes in file.
-        """
-        data_array = np.load(self.filename(date=date), mmap_mode="r")
-        return data_array[
-            self.grid.subdomain[0] : self.grid.subdomain[1],
-            self.grid.subdomain[2] : self.grid.subdomain[3],
-            term,
-            member,
-        ]
+    array = load_data(settings.dataset_name, param, date, terms, member)
 
-    def exist(self, date: dt.datetime) -> bool:
-        flist = self.filename(date=date)
-        return flist.exists()
+    # Extend dimension to match 3D (level dimension)
+    if len(array.shape) != 4:
+        array = np.expand_dims(array, axis=-1)
+    array = np.transpose(array, axes=[2, 0, 1, 3])  # shape = (steps, lvl, x, y)
 
+    if standardize:
+        array = (array - means) / std
 
-@dataclass(slots=True)
-class PoesySettings:
-    term: dict
-    num_input_steps: int  # = 2  # Number of input timesteps
-    num_output_steps: int  # = 1  # Number of output timesteps (= 0 for inference)
-    num_inference_pred_steps: int = (
-        0  # 0 in training config ; else used to provide future information about forcings
-    )
-    standardize: bool = False
-    members: Tuple[int] = (0,)
+    # Define which value is considered invalid
+    tensor_data = torch.from_numpy(array)
 
-    @property
-    def num_total_steps(self) -> int:
-        """
-        Total number of timesteps
-        for one sample.
-        """
-        # Nb of step in one sample
-        return self.num_input_steps + self.num_output_steps
+    return tensor_data
 
 
 @dataclass(slots=True)
 class Sample:
-    # Describe a sample
-    # TODO consider members
+    """Describes a sample"""
+
     member: int
     date: dt.datetime
+    settings: Settings
     input_terms: Tuple[float]
     output_terms: Tuple[float]
+    stats: Stats
+    grid: Grid
+    params: List[Param]
 
     # Term wrt to the date {date}. Gives validity
     terms: Tuple[float] = field(init=False)
 
     def __post_init__(self):
         self.terms = self.input_terms + self.output_terms
 
-    def is_valid(self, param_list: List) -> bool:
+    def is_valid(self) -> bool:
         """
         Check that all the files necessary for this samples exists.
 
         Args:
             param_list (List): List of parameters
         Returns:
-            Boolean:  Whether the sample exist or not
+            Boolean:  Whether the sample exists or not
         """
-        for param in param_list:
-            if not param.exist(self.date):
+        for param in self.params:
+            if not exists(self.settings.dataset_name, param, self.date):
                 return False
 
         return True
 
+    def generate_forcings(
```

Comment:
I forgot to do it,  but I think we can withdraw this method from sample and make an independant function, such as in Titan

Reply:
done